### PR TITLE
Saving external data for large ONNX models

### LINKED ISSUE #254: Saving external data for > 2GB models
### Who can help?

@philschmid 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```python
from optimum.onnxruntime import ORTModelForCausalLM

model_ckpt = "bigscience/bloom-350m"
model = ORTModelForCausalLM.from_pretrained(model_ckpt, use_auth_token=True, from_transformers=True)
model.save_pretrained("output/") # saves only model.onnx without the rest of external data
```

### Expected behavior
The script above should save all the following files:
```shell
model.onnx         onnx__MatMul_9335  onnx__MatMul_9561  onnx__MatMul_9815                                     transformer.h.1.mlp.dense_h_to_4h.bias
onnx__MatMul_9086  onnx__MatMul_9336  onnx__MatMul_9566  onnx__MatMul_9816                                     transformer.h.1.self_attention.query_key_value.bias
onnx__MatMul_9111  onnx__MatMul_9337  onnx__MatMul_9591  onnx__MatMul_9817                                     transformer.h.20.mlp.dense_h_to_4h.bias
onnx__MatMul_9112  onnx__MatMul_9342  onnx__MatMul_9592  onnx__MatMul_9822                                     transformer.h.20.self_attention.query_key_value.bias
onnx__MatMul_9113  onnx__MatMul_9367  onnx__MatMul_9593  onnx__MatMul_9847                                     transformer.h.21.mlp.dense_h_to_4h.bias
onnx__MatMul_9118  onnx__MatMul_9368  onnx__MatMul_9598  onnx__MatMul_9848                                     transformer.h.21.self_attention.query_key_value.bias
onnx__MatMul_9143  onnx__MatMul_9369  onnx__MatMul_9623  onnx__MatMul_9849                                     transformer.h.22.mlp.dense_h_to_4h.bias
onnx__MatMul_9144  onnx__MatMul_9374  onnx__MatMul_9624  onnx__MatMul_9850                                     transformer.h.22.self_attention.query_key_value.bias
onnx__MatMul_9145  onnx__MatMul_9399  onnx__MatMul_9625  transformer.h.0.mlp.dense_h_to_4h.bias                transformer.h.23.mlp.dense_h_to_4h.bias
onnx__MatMul_9150  onnx__MatMul_9400  onnx__MatMul_9630  transformer.h.0.self_attention.query_key_value.bias   transformer.h.23.self_attention.query_key_value.bias
onnx__MatMul_9175  onnx__MatMul_9401  onnx__MatMul_9655  transformer.h.10.mlp.dense_h_to_4h.bias               transformer.h.2.mlp.dense_h_to_4h.bias
onnx__MatMul_9176  onnx__MatMul_9406  onnx__MatMul_9656  transformer.h.10.self_attention.query_key_value.bias  transformer.h.2.self_attention.query_key_value.bias
onnx__MatMul_9177  onnx__MatMul_9431  onnx__MatMul_9657  transformer.h.11.mlp.dense_h_to_4h.bias               transformer.h.3.mlp.dense_h_to_4h.bias
onnx__MatMul_9182  onnx__MatMul_9432  onnx__MatMul_9662  transformer.h.11.self_attention.query_key_value.bias  transformer.h.3.self_attention.query_key_value.bias
onnx__MatMul_9207  onnx__MatMul_9433  onnx__MatMul_9687  transformer.h.12.mlp.dense_h_to_4h.bias               transformer.h.4.mlp.dense_h_to_4h.bias
onnx__MatMul_9208  onnx__MatMul_9438  onnx__MatMul_9688  transformer.h.12.self_attention.query_key_value.bias  transformer.h.4.self_attention.query_key_value.bias
onnx__MatMul_9209  onnx__MatMul_9463  onnx__MatMul_9689  transformer.h.13.mlp.dense_h_to_4h.bias               transformer.h.5.mlp.dense_h_to_4h.bias
onnx__MatMul_9214  onnx__MatMul_9464  onnx__MatMul_9694  transformer.h.13.self_attention.query_key_value.bias  transformer.h.5.self_attention.query_key_value.bias
onnx__MatMul_9239  onnx__MatMul_9465  onnx__MatMul_9719  transformer.h.14.mlp.dense_h_to_4h.bias               transformer.h.6.mlp.dense_h_to_4h.bias
onnx__MatMul_9240  onnx__MatMul_9470  onnx__MatMul_9720  transformer.h.14.self_attention.query_key_value.bias  transformer.h.6.self_attention.query_key_value.bias
onnx__MatMul_9241  onnx__MatMul_9495  onnx__MatMul_9721  transformer.h.15.mlp.dense_h_to_4h.bias               transformer.h.7.mlp.dense_h_to_4h.bias
onnx__MatMul_9246  onnx__MatMul_9496  onnx__MatMul_9726  transformer.h.15.self_attention.query_key_value.bias  transformer.h.7.self_attention.query_key_value.bias
onnx__MatMul_9271  onnx__MatMul_9497  onnx__MatMul_9751  transformer.h.16.mlp.dense_h_to_4h.bias               transformer.h.8.mlp.dense_h_to_4h.bias
onnx__MatMul_9272  onnx__MatMul_9502  onnx__MatMul_9752  transformer.h.16.self_attention.query_key_value.bias  transformer.h.8.self_attention.query_key_value.bias
onnx__MatMul_9273  onnx__MatMul_9527  onnx__MatMul_9753  transformer.h.17.mlp.dense_h_to_4h.bias               transformer.h.9.mlp.dense_h_to_4h.bias
onnx__MatMul_9278  onnx__MatMul_9528  onnx__MatMul_9758  transformer.h.17.self_attention.query_key_value.bias  transformer.h.9.self_attention.query_key_value.bias
onnx__MatMul_9303  onnx__MatMul_9529  onnx__MatMul_9783  transformer.h.18.mlp.dense_h_to_4h.bias               transformer.word_embeddings.weight
onnx__MatMul_9304  onnx__MatMul_9534  onnx__MatMul_9784  transformer.h.18.self_attention.query_key_value.bias
onnx__MatMul_9305  onnx__MatMul_9559  onnx__MatMul_9785  transformer.h.19.mlp.dense_h_to_4h.bias
onnx__MatMul_9310  onnx__MatMul_9560  onnx__MatMul_9790  transformer.h.19.self_attention.query_key_value.bias
```

### PR DESCRIPTION
**Pull request by NouamaneTazi**
# What does this PR do?

Fixes #254 and https://github.com/huggingface/optimum/issues/377

We can now load and save ORT models that have external data 🚀

### CODE REVIEW

**Comment by NouamaneTazi** [comment]
With the latest commit, we're now able to do:
```python
model = ORTModelForCausalLM.from_pretrained(
    model_ckpt,
    use_auth_token=True,
    from_transformers=True,
    cache="model_cache",
    onnx_cache_dir="./onnx_cache",  # saves ONNX model with external data if large model to "./onnx_cache"
)

model = ORTModelForCausalLM.from_pretrained(
    model_ckpt,
    use_auth_token=True,
    from_transformers=True,
    cache="model_cache"  # like previous behaviour where `onnx_cache_dir`= `cache`
)
```

And `model.save_pretrained(save_path)` would just copy files from `onnx_cache_dir` to the provided `save_path`

**Comment by NouamaneTazi** [comment]
The following should be working now
```python
# load small ONNX model
model = ORTModelForCausalLM.from_pretrained("nouamanetazi/bloom-small-testing-onnx", use_auth_token=True)
# load large ONNX model (>2GB)
model = ORTModelForCausalLM.from_pretrained("nouamanetazi/bloom-350m-onnx-folder", use_auth_token=True, onnx_folder="onnx")
```

File: optimum/onnxruntime/modeling_ort.py [hunk 1]
```diff
@@ -182,11 +185,16 @@ def _from_pretrained(
...
                 repository or directory.
+            onnx_folder(`str`):
+                Folder path in repo where the onnx model with external data is located. If not specified, the onnx model is assumed to be in the root of the repo.
+                Note: for large models (>2GB), this argument must be specified to load the model's external data.
             kwargs (`Dict`, *optional*):
                 kwargs will be passed to the model during initialization
         """
         config_dict = kwargs.pop("config", {})
         model_file_name = file_name if file_name is not None else ONNX_WEIGHTS_NAME
+        #TODO: if we can find a good regex for onnx's external data, we don't need to specify the onnx_folder
```
Comment (philschmid) [inline optimum/onnxruntime/modeling_ort.py:196]: @mfuntowicz do you know some?

**Review by philschmid — CHANGES_REQUESTED** [review]
Can you make sure tests are passing and style is correct?

**Comment by NouamaneTazi** [comment]
We can now save/load large `ORTModelForSeq2SeqLM`
```
from pathlib import Path
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForSeq2SeqLM

model_ckpt = "facebook/mbart-large-en-ro"
save_path = Path(f"saved_model/{model_ckpt}")
save_path.mkdir(parents=True, exist_ok=True)

tokenizer = AutoTokenizer.from_pretrained(model_ckpt, use_auth_token=True)
model = ORTModelForSeq2SeqLM.from_pretrained(
    model_ckpt,
    use_auth_token=True,
    from_transformers=True,
)

# # save to local folder
model.save_pretrained(save_path / "onnx")
tokenizer.save_pretrained(save_path)

```

File: optimum/onnxruntime/utils.py [hunk 2]
```diff
@@ -271,3 +271,23 @@ class ORTQuantizableOperator(Enum):
     Resize = "Resize"
     AveragePool = "AveragePool"
     Concat = "Concat"
+
+
+def _get_external_data_paths(src_file_names, dst_file_names):
```
Comment (fxmarty) [inline optimum/onnxruntime/utils.py:276]: Could you document (even just a sentence) this method? i think it makes it easier for people not super familiar with the code / ONNX / ORT to understand it.

File: optimum/onnxruntime/utils.py [hunk 3]
```diff
@@ -271,3 +271,23 @@ class ORTQuantizableOperator(Enum):
...
+    import onnx
```
Comment (fxmarty) [inline optimum/onnxruntime/utils.py:277]: I can see the reason to put imports here if it is only used in this method - but is this standard? Actually never sure about it

**Comment by fxmarty** [comment]
Awesome! I think it would be great to add tests, essentially that saving / reloading works well, in the encoder-only/encoder-decoder cases.

File: optimum/onnxruntime/utils.py (hunk 3 above)
Comment (michaelbenayoun) [inline optimum/onnxruntime/utils.py:277]: Let's make the imports at the top of the file

File: optimum/onnxruntime/utils.py [hunk 4]
```diff
@@ -271,3 +271,23 @@ class ORTQuantizableOperator(Enum):
...
+    from onnx.external_data_helper import ExternalDataInfo, _get_initializer_tensors
+
+    model_paths = src_file_names.copy()
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/utils.py:280]: It's a list of Paths so I don't think it copies anything here, we might as well start from a new empty list, and fill it.

File: optimum/onnxruntime/utils.py (hunk 2 above)
Comment (michaelbenayoun) [inline optimum/onnxruntime/utils.py:276]: I agree that we should change the naming to make it clear about the types.
Also add type annotation:
```suggestion
def _get_external_data_paths(src_paths: List[Path], dst_file_names: List[str]) -> Tuple[List[Path], List[str]]:
```

File: optimum/onnxruntime/utils.py (hunk 4 above)
Comment (NouamaneTazi) [reply optimum/onnxruntime/utils.py:280]: I checked without the `copy()` and the extending does modify the list `src_file_names` unfortunately :/

**Comment by NouamaneTazi** [comment]
For the tests, it would be cool if we could enforce saving a small model in external data format. I tried looking quickly for a way, but there doesn't seem to be an easy way to bypass the 2GB protobuf file limit.
Will try to add tests once I have time @fxmarty

File: optimum/onnxruntime/utils.py (hunk 4 above)
Comment (michaelbenayoun) [reply optimum/onnxruntime/utils.py]: Yes, my point is that you can create an empty list

```python
model_paths = []
```

And fill it as you go?

My point here is that `src_files_names[0]` will be the same instance of `model_paths[0]`.

Comment (NouamaneTazi) [reply optimum/onnxruntime/utils.py]: I'm sorry I don't quite get it. I only use `model_paths` to iterate over the inputted `src_files_names`. Then I keep on filling `src_files_names`
```
    model_paths = []
    for model_path in model_paths:
        # load model graph
        model = onnx.load(str(model_path), load_external_data=False)
        # filter out tensors that are not external data
        model_tensors = _get_initializer_tensors(model)
        model_tensors_ext = [
            ExternalDataInfo(tensor).location
            for tensor in model_tensors
            if tensor.HasField("data_location") and tensor.data_location == onnx.TensorProto.EXTERNAL
        ]
        src_paths.extend([model_path.parent / tensor_name for tensor_name in model_tensors_ext])
        dst_file_names.extend(model_tensors_ext)
    return src_paths, dst_file_names
```
So this shouldn't work

**Comment by mht-sharma** [comment]
> For the tests, it would be cool if we could enforce saving a small model in external data format. I tried looking quickly for a way, but there doesn't seem to be an easy way to bypass the 2GB protobuf file limit.
> Will try to add tests once I have time @fxmarty

You could use the following api to convert a small model to external data. [converting-an-onnx-model-to-external-data](https://github.com/onnx/onnx/blob/main/docs/ExternalData.md#converting-an-onnx-model-to-external-data
)

The size threshold has to be low so that it can create the files.

**Comment by mht-sharma** [comment]
@NouamaneTazi The PR would require a small change to handle one more use-case for the `modelling_seq2seq`  and `modelling_decoder`.  Taking an example of `Seq2Seq` class, it generates 3 different models, `encoder.onnx`, `decoder.onnx`, `decoder_with_past.onnx`. These are generated in the root folder right now. In case there are external files, there is a chance of an overwrite if they have the same names. See: [26983](https://discuss.huggingface.co/t/fail-onnxruntimeerror-1-fail-deserialize-tensor-onnx/26983/4)

Possible fix is to save them in folders like, `encoder/encoder.onnx`, `decoder/decoder.onnx' etc. The same change would be required in the exporters.

**Comment by fxmarty** [comment]
We should probably do the same in exporters actually

File: optimum/onnxruntime/utils.py [hunk 5]
```diff
@@ -270,3 +272,23 @@ class ORTQuantizableOperator(Enum):
...
     Concat = "Concat"
+
+
+def _get_external_data_paths(src_paths: List[Path], dst_file_names: List[str]) -> Tuple[List[Path], List[str]]:
+    """
+    Get external data paths from the model and add them to the list of files to copy.
+    """
+    model_paths = src_paths.copy()
+    for model_path in model_paths:
+        # load model graph
```
Comment (michaelbenayoun) [inline optimum/onnxruntime/utils.py:283]: Not needed
```suggestion
```

**Review by michaelbenayoun — APPROVED** [review]
LGTM!

**Comment by fxmarty** [comment]
@NouamaneTazi Why not use actual >2GB models, initialized and saved random from transformers (so no download time)?

**Comment by NouamaneTazi** [comment]
@fxmarty Yes definitely! I can use a randomly intialized model, but It seems there's no exposed API to load for example `ORTModelForSequenceClassification` from a `BertForSequenceClassification` instance?

**Comment by fxmarty** [comment]
> @fxmarty Yes definitely! I can use a randomly intialized model, but It seems there's no exposed API to load for example `ORTModelForSequenceClassification` from a `BertForSequenceClassification` instance?

You can do `save_pretrained()` on the PretrainedModel, and then `from_pretrained` then from a local folder using ORTModel.

**Comment by PoodleWang** [comment]
@NouamaneTazi 
from pathlib import Path
from transformers import AutoTokenizer
from optimum.onnxruntime import ORTModelForSeq2SeqLM

model_ckpt = "facebook/mbart-large-en-ro"
save_path = Path(f"saved_model/{model_ckpt}")
save_path.mkdir(parents=True, exist_ok=True)

tokenizer = AutoTokenizer.from_pretrained(model_ckpt)
model = ORTModelForSeq2SeqLM.from_pretrained(
    model_ckpt,
    from_transformers=True,
)
model.save_pretrained(save_path / "onnx")
tokenizer.save_pretrained(save_path)

Log:
<Trial 2015437 worker_0> genius $ python3 /opt/tiger/genius/tensorrt/load.py
2022-12-15 09:13:01.539465: I tensorflow/core/util/util.cc:169] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
/home/tiger/.local/lib/python3.7/site-packages/transformers/models/mbart/modeling_mbart.py:239: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if attn_weights.size() != (bsz * self.num_heads, tgt_len, src_len):
/home/tiger/.local/lib/python3.7/site-packages/transformers/models/mbart/modeling_mbart.py:246: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if attention_mask.size() != (bsz, 1, tgt_len, src_len):
/home/tiger/.local/lib/python3.7/site-packages/transformers/models/mbart/modeling_mbart.py:278: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if attn_output.size() != (bsz * self.num_heads, tgt_len, self.head_dim):
/home/tiger/.local/lib/python3.7/site-packages/transformers/models/mbart/modeling_mbart.py:912: TracerWarning: Converting a tensor to a Python boolean might cause the trace to be incorrect. We can't record the data flow of Python values, so this value will be treated as a constant in the future. This means that the trace might not generalize to other inputs!
  if input_shape[-1] > 1:
/home/tiger/.local/lib/python3.7/site-packages/transformers/models/mbart/modeling_mbart.py:100: TracerWarning: torch.tensor results are registered as constants in the trace. You can safely ignore this warning if you use this function to create tensors out of constant variables that would be the same every time you call this function. In any other case, this might cause the trace to be incorrect.
  mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min))
Traceback (most recent call last):
  File "/opt/tiger/genius/tensorrt/load.py", line 55, in <module>
    from_transformers=True,
  File "/usr/local/lib/python3.7/dist-packages/optimum/onnxruntime/modeling_ort.py", line 280, in from_pretrained
    **kwargs,
  File "/usr/local/lib/python3.7/dist-packages/optimum/modeling_base.py", line 263, in from_pretrained
    **model_kwargs,
  File "/usr/local/lib/python3.7/dist-packages/optimum/onnxruntime/modeling_seq2seq.py", line 597, in _from_transformers
    output=save_dir.joinpath(ONNX_DECODER_NAME),
  File "/home/tiger/.local/lib/python3.7/site-packages/transformers/onnx/convert.py", line 353, in export
    return export_pytorch(preprocessor, model, config, opset, output, tokenizer=tokenizer, device=device)
  File "/home/tiger/.local/lib/python3.7/site-packages/transformers/onnx/convert.py", line 204, in export_pytorch
    raise err
  File "/home/tiger/.local/lib/python3.7/site-packages/transformers/onnx/convert.py", line 189, in export_pytorch
    opset_version=opset,
  File "/home/tiger/.local/lib/python3.7/site-packages/torch/onnx/__init__.py", line 280, in export
    custom_opsets, enable_onnx_checker, use_external_data_format)
  File "/home/tiger/.local/lib/python3.7/site-packages/torch/onnx/utils.py", line 94, in export
    use_external_data_format=use_external_data_format)
  File "/home/tiger/.local/lib/python3.7/site-packages/torch/onnx/utils.py", line 706, in _export
    val_add_node_names, val_use_external_data_format, model_file_location)
RuntimeError: Exporting model exceed maximum protobuf size of 2GB. Please call torch.onnx.export with use_external_data_format=True.

** Environment
optimum-1.5.1

**Comment by PoodleWang** [comment]
I have similar issue mentioned here: https://github.com/huggingface/optimum/issues/589#issuecomment-1352465502

**Comment by NouamaneTazi** [comment]
Migrated this PR to https://github.com/huggingface/optimum/pull/586

**Pull request closed without merging**